File: Packages/StackEngineCore/Tests/StackEngineCoreTests/Transform2DTests.swift
import XCTest
import simd
@testable import StackEngineCore

final class Transform2DTests: XCTestCase {
    func testIdentityMapsPointToItself() {
        let p = Transform2D.identity.apply(3, 4)
        XCTAssertEqual(p.x, 3, accuracy: 1e-6)
        XCTAssertEqual(p.y, 4, accuracy: 1e-6)
    }

    func testSimilarityScalesAndTranslates() {
        let t = Transform2D.similarity(scale: 2, rotation: 0, tx: 1, ty: -1)
        let p = t.apply(3, 4)
        XCTAssertEqual(p.x, 7, accuracy: 1e-6)   // 2·3 + 1
        XCTAssertEqual(p.y, 7, accuracy: 1e-6)   // 2·4 − 1
    }

    func testSimilarityRotatesNinetyDegrees() {
        let t = Transform2D.similarity(scale: 1, rotation: .pi / 2, tx: 0, ty: 0)
        let p = t.apply(1, 0)
        XCTAssertEqual(p.x, 0, accuracy: 1e-6)    // (1,0) rotated +90° → (0,1)
        XCTAssertEqual(p.y, 1, accuracy: 1e-6)
    }

    func testComposedAppliesRightHandSideFirst() {
        let scale = Transform2D.similarity(scale: 2, rotation: 0, tx: 0, ty: 0)
        let shift = Transform2D.similarity(scale: 1, rotation: 0, tx: 3, ty: -1)
        // scale ∘ shift: p → scale(shift(p)) = 2·(p + (3,−1))
        let p = scale.composed(with: shift).apply(1, 1)
        XCTAssertEqual(p.x, 8, accuracy: 1e-5)
        XCTAssertEqual(p.y, 0, accuracy: 1e-5)
    }

    func testComposedWithIdentityIsUnchanged() {
        let t = Transform2D.similarity(scale: 1.04, rotation: 0.02, tx: 2, ty: -1)
        // exact ==: multiplying by exact 1.0/0.0 is lossless in IEEE — don't weaken to tolerances.
        XCTAssertEqual(t.composed(with: .identity), t)
        XCTAssertEqual(Transform2D.identity.composed(with: t), t)
    }

    func testInverseRoundTripsToIdentity() {
        let t = Transform2D.similarity(scale: 1.04, rotation: 0.02, tx: 2, ty: -1)
        let id = t.composed(with: t.inverse)
        XCTAssertEqual(id.a, 1, accuracy: 1e-5)
        XCTAssertEqual(id.b, 0, accuracy: 1e-5)
        XCTAssertEqual(id.c, 0, accuracy: 1e-5)
        XCTAssertEqual(id.d, 1, accuracy: 1e-5)
        XCTAssertEqual(id.tx, 0, accuracy: 1e-4)
        XCTAssertEqual(id.ty, 0, accuracy: 1e-4)
    }
}

File: android/stackengine/src/test/kotlin/com/jdpneto/stackengine/Transform2DTests.kt
package com.jdpneto.stackengine

import kotlin.math.PI
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class Transform2DTests {

    @Test
    fun testIdentityMapsPointToItself() {
        val (px, py) = Transform2D.identity.apply(3f, 4f)
        assertEquals(3f, px, absoluteTolerance = 1e-6f)
        assertEquals(4f, py, absoluteTolerance = 1e-6f)
    }

    @Test
    fun testSimilarityScalesAndTranslates() {
        val t = Transform2D.similarity(scale = 2f, rotation = 0f, tx = 1f, ty = -1f)
        val (px, py) = t.apply(3f, 4f)
        assertEquals(7f, px, absoluteTolerance = 1e-6f)   // 2·3 + 1
        assertEquals(7f, py, absoluteTolerance = 1e-6f)   // 2·4 − 1
    }

    @Test
    fun testSimilarityRotatesNinetyDegrees() {
        val t = Transform2D.similarity(scale = 1f, rotation = (PI / 2).toFloat(), tx = 0f, ty = 0f)
        val (px, py) = t.apply(1f, 0f)
        assertEquals(0f, px, absoluteTolerance = 1e-6f)    // (1,0) rotated +90° → (0,1)
        assertEquals(1f, py, absoluteTolerance = 1e-6f)
    }

    @Test
    fun testComposedAppliesRightHandSideFirst() {
        val scale = Transform2D.similarity(scale = 2f, rotation = 0f, tx = 0f, ty = 0f)
        val shift = Transform2D.similarity(scale = 1f, rotation = 0f, tx = 3f, ty = -1f)
        // scale ∘ shift: p → scale(shift(p)) = 2·(p + (3,−1))
        val (px, py) = scale.composed(with = shift).apply(1f, 1f)
        assertEquals(8f, px, absoluteTolerance = 1e-5f)
        assertEquals(0f, py, absoluteTolerance = 1e-5f)
    }

    @Test
    fun testComposedWithIdentityIsUnchanged() {
        val t = Transform2D.similarity(scale = 1.04f, rotation = 0.02f, tx = 2f, ty = -1f)
        // exact ==: multiplying by exact 1.0/0.0 is lossless in IEEE — don't weaken to tolerances.
        assertEquals(t, t.composed(with = Transform2D.identity))
        assertEquals(t, Transform2D.identity.composed(with = t))
    }

    @Test
    fun testInverseRoundTripsToIdentity() {
        val t = Transform2D.similarity(scale = 1.04f, rotation = 0.02f, tx = 2f, ty = -1f)
        val id = t.composed(with = t.inverse)
        assertEquals(1f, id.a, absoluteTolerance = 1e-5f)
        assertEquals(0f, id.b, absoluteTolerance = 1e-5f)
        assertEquals(0f, id.c, absoluteTolerance = 1e-5f)
        assertEquals(1f, id.d, absoluteTolerance = 1e-5f)
        assertEquals(0f, id.tx, absoluteTolerance = 1e-4f)
        assertEquals(0f, id.ty, absoluteTolerance = 1e-4f)
    }
}

// Re-export tolerance helper so it looks like Swift's accuracy: parameter
private fun assertEquals(expected: Float, actual: Float, absoluteTolerance: Float) {
    assertTrue(
        kotlin.math.abs(expected - actual) <= absoluteTolerance,
        "expected $expected but got $actual (tolerance $absoluteTolerance)"
    )
}
